### PR TITLE
Feature/custom collection

### LINKED ISSUE #223: Allow users to add a custom collection in config
### Summary

Users can add a custom collection id in the h2oGPTe action config

### Motivation

Allowing a custom collection_id lets teams point the h2oGPTe Action at the exact knowledge base for the agent. It improves reproducibility (consistent context across runs) and avoids rebuilding or re-indexing data for each pipeline.

### Proposed Solution

Add a ‎`collection_id` input to the h2oGPTe GitHub Action. For example:

```
# .github/workflows/h2ogpte.yml
with:
  policy_profile: enterprise-default
  collection_id: kb-fincrime-aml
```

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Sanktrip**
Closes #223 and potentially #58

### CODE REVIEW

File: docs/USAGE.md [hunk 1]
```diff
@@ -67,6 +67,10 @@ Maintains comprehensive documentation standards across your repository. This wor
 
 Ensures your code changes never ship without proper test coverage. The agent reviews each pull request, identifies untested code paths, and automatically creates comprehensive test files that cover edge cases and follow your project's testing conventions. It only adds tests - never modifies existing code.
 
+### 📂 Custom Collections ([h2ogpte_custom_collection.yaml](../examples/custom_workflows/h2ogpte_custom_collection.yaml))
+
+Creates a new collection containing the settings and file contents of the provided collection. The agent now has an exact and standardised knowledge base, improving reproducibility.
+
```
Comment (MillenniumForce) [inline docs/USAGE.md]: It's not really clear how the agent should use the custom collection. Probably need to add an additional instruction to the system prompt to ask the agent to read the collection for any existing files.

File: examples/custom_workflows/h2ogpte_custom_collection.yaml
Comment (MillenniumForce) [inline examples/custom_workflows/h2ogpte_custom_collection.yaml]: I think it's overkill to add another custom workflow just for custom collections. Safe to delete in my opinion.

File: src/core/services/h2ogpte/h2ogpte.ts [hunk 2]
```diff
@@ -416,3 +415,306 @@ export async function deleteCollection(
...
   );
 }
+
+/**
+ * Validates if a collection exists and is accessible
+ * @param collectionId - The ID of the collection to validate
+ * @returns Promise<boolean> - True if collection is valid and accessible
+ * @throws Error if the API request fails
+ */
+export async function isValidCollection(
```
Comment (MillenniumForce) [inline src/core/services/h2ogpte/h2ogpte.ts]: Same as in the previous review, `src/core/services/h2ogpte/h2ogpte.ts` is only meant for low-level api requests. Maybe change this function name to `getCollection` so its explicit?

File: src/core/services/h2ogpte/h2ogpte.ts [hunk 3]
```diff
@@ -416,3 +415,306 @@ export async function deleteCollection(
...
+}
+
+/**
+ * Duplicates a collection by copying settings, chat configuration, and documents
+ * @param sourceCollectionId - The ID of the collection to duplicate from
+ * @param targetCollectionId - The ID of the collection to duplicate to
+ * @returns Promise<void>
+ * @throws Error if duplication fails at any step
+ */
+export async function duplicateCollection(
```
Comment (MillenniumForce) [inline src/core/services/h2ogpte/h2ogpte.ts]: Same issue with this function see: https://github.com/h2oai/h2ogpte-action/pull/252/changes#r2752545622

I think this should be moved to `src/core/services/h2ogpte/utils.ts`

File: tests/h2ogpte/collection-id.test.ts
Comment (MillenniumForce) [inline tests/h2ogpte/collection-id.test.ts]: I'm not sure these unit tests add any value. Since they just mock all of the responses I think we're better off doing integration tests.

File: action.yml [hunk 4]
```diff
@@ -45,9 +45,8 @@ inputs:
   prompt:
     description: Prompt for custom workflow. String injection options are {{repoName}}, {{idNumber}} and {{eventsText}}, see README.md for more details.
     required: false
-  slash_commands:
-    description: JSON string defining slash commands. Each command should have a name and prompt field.
-    required: false
```
Comment (MillenniumForce) [inline action.yml]: Keep slash commands

File: docs/USAGE.md (hunk 1 above)
Comment (MillenniumForce) [reply docs/USAGE.md]: As discussed in person, we should instead instruct the user how to prompt the agent to use their custom collection.

This could be done in two ways:
1. User adds a prompt when they ask h2ogpte
2. User adds instruction into the prompt when creating a custom workflow (should link to the custom workflows section)

File: src/core/services/h2ogpte/h2ogpte.ts [hunk 5]
```diff
@@ -416,3 +415,266 @@ export async function deleteCollection(
...
+    const errorText = await response.text();
+    throw new Error(
+      `Failed to get collection documents: ${response.status} ${response.statusText} - ${errorText}`,
+    );
+  }
+  const data = (await response.json()) as types.Document[];
+  return data;
+}
+
+export async function addDocumentsToCollection(
```
Comment (MillenniumForce) [inline src/core/services/h2ogpte/h2ogpte.ts]: no doc string, personally I think we should just get rid of all doc strings. Code should be self documenting. Up to what you prefer but we should be consistent.

File: src/core/services/h2ogpte/utils.ts [hunk 6]
```diff
@@ -81,3 +96,55 @@ export function parseH2ogpteConfig(): H2ogpteConfig {
...
+
+  // Get source collection documents
+  const documents = (await getCollectionDocumentsData(
+    sourceCollectionId,
+  )) as Document[];
+  // Add documents to target collection
+  await addDocumentsToCollection(
+    targetCollectionId,
+    documents.map((doc) => doc.id),
+  );
```
Comment (MillenniumForce) [inline src/core/services/h2ogpte/utils.ts]: Personally, I think `addDocumentsToCollection` function should be `addDocumentToCollection` (i.e. process singular document) and then apply the `Promise.all` in this function

**Review by MillenniumForce — APPROVED** [review]
Thanks, approved

File: docs/CONFIGURATION.md [hunk 7]
```diff
@@ -4,14 +4,15 @@ The h2oGPTe GitHub Action supports several configuration options to customize th
...
+| **Agent Total Timeout (`agent_total_timeout`)** | `3600`       | Any positive integer (in seconds)                                                                                            | Set the maximum time (in seconds) the agent can run before timing out. Default is 3600 seconds (1 hour). Invalid or negative values will use the default.                                                                                                                                     |
+| **Custom Collection (`collection_id`)**         | -            | String                                                                                                                       | Creates a new duplicate collection with the same settings and files as the specified collection. The original collection remains unchanged. New files from the PR/issue/comment will be added to the duplicated collection. If not specified, a new empty collection is created for each run. |
 
-## Configuration Example
+## ✅ 3. Configuration Example
```
[the remark this answers is not in the document]
Comment (MillenniumForce) [reply docs/CONFIGURATION.md]: Emoji is still present

**Review by MillenniumForce — CHANGES_REQUESTED** [review]

File: src/core/services/h2ogpte/h2ogpte.ts [hunk 8]
```diff
@@ -449,6 +482,38 @@ export async function getCollectionSettings(
...
+    throw new Error(
+      `Failed to get chat settings: ${response.status} ${response.statusText} - ${errorText}`,
+    );
+  }
+  const data = (await response.json()) as types.ChatSettings;
+  return data;
+}
+
 export async function setCollectionSettings(
   collectionId: string,
```
[the remark this answers is not in the document]
Comment (Sanktrip) [reply src/core/services/h2ogpte/h2ogpte.ts]: Nope, they both overwrite the existing settings completely. I can rename them accordingly.

File: docs/CONFIGURATION.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (MillenniumForce) [reply docs/CONFIGURATION.md]: Header still isn't fixed, should be `## Configuration Example`
